pkgs/jnigen/doc/lifecycle.md

```diff
@@ -0,0 +1,150 @@
+the Dart GC collects them, the underlying Java reference is released.
+
+This approach works well for application code where a large number of references
+aren't created repeatedly. However, it's not ideal for packages because the
+usage patterns are unpredictable. This can lead to exceeding the global
+reference limit and application crashes.
+
+### Eagerly releasing references manually (recommended for packages)
+
+Instead of waiting for Dart GC to release the JNI global references,
+`.release()` can be called on the `JObject`s.
+
+```dart
+// Construct the object.
+final hello = 'Hello'.toJString();
+// Use it.
+print(hello);
+// Eagerly release it!
+hello.release();
+```
+
+To make releasing easier, we can use
+[`Arena`](https://pub.dev/documentation/ffi/latest/ffi/Arena-class.html)s.
+First, create an `Arena` via
+[`using`](https://pub.dev/documentation/ffi/latest/ffi/using.html). Then
+register the object to be released at the end of the callback.
+
+```dart
+using((arena) {
+  final hello = 'Hello'.toJString()..releasedBy(arena);
+  final world = 'World'.toJString()..releasedBy(arena);
+  print(hello);
+  print(world);
+});
+// Both `hello` and `world` are now released.
+```
+
+### Lifecycle of interfaces implemented in Dart
+
+Java interfaces can be implemented in Dart using `.implement`. The Dart closure
+implementations must remain accessible as long as the created Java object
+exists. Ownership of these Dart objects is transferred to Java. When the Java GC
+reclaims the implemented object, a message is sent to Dart to remove the
+corresponding closures.
+
+> [!WARNING]
+>
+> One could create cycles between Dart and Java GC's when implementing
+> interfaces. For example consider the following:
+>
+> ```dart
+> final foo = Foo();
+> foo.bar = Bar.implement($Bar(
+>   f: () {
+>    return foo;
+>   }
+> ));
+> ```
+>
+> In Dart, closure `f` holds a reference to `foo`, preventing its garbage
+> collection. In Java, `foo` holds a reference to `bar`, preventing its release
+> and thus preventing the closure's removal. This creates a cycle.
+>
+> ```mermaid
+> graph TD;
+>    f-->|holds in Dart|foo;
+>    foo-->|holds in Java|bar;
+>    bar-->|holds in Java|f;
+> ```
+
+To prevent cycles, use
+[`WeakReference`](https://api.dart.dev/dart-core/WeakReference-class.html)s.
+
+```dart
+final weakFoo = WeakReference(foo);
+foo.bar = Bar.implement($Bar(
+  f: () {
+    final foo = weakFoo.target;
+    if (foo == null) {
+      throw StateError();
+    }
+    return foo;
+  }
+));
+```
+
+The weak reference breaks the cycle, allowing the GCs to collect the objects.
+
+```mermaid
+graph TD;
+   f-.->|holds weakly in Dart|foo;
+   foo-->|holds in Java|bar;
+   bar-->|holds in Java|f;
+```
+
+> [!WARNING]
+>
+> Be careful about the closures accidentally overcapturing. To prevent
+> overcapturing, implement your logic in a separate function or create a class
+> that implements `$Bar`.
```

Comment:
Can you give an example here by modifying the above code to make `f` a separate function?